Reword README

fulldecent opened the pull request:
This borrows HEAVILY from https://github.com/apple/swift-evolution/blob/master/README.md and https://github.com/apple/swift-evolution/blob/master/process.md which is a mature and well-respected process from Apple. The difference is that they are a community with parties that set direction; but nobody/everybody sets direction for the Ethereum community.

Overview of changes proposed in this PR:

* Sets project goal
* Achieving "Final" status in this repository only represents that a proposal has been reviewed for technical accuracy. @MicahZoltu @Arachnid  Did I finally get that right?
* Endorses EM as the place to discuss stuff BEFORE making a proposal

# Work plan

- [ ] Merge this PR
- [ ] Disable the GitHub Issues feature on this repository, every "issue" I see here is actually something that should be discussed on EM
- [x] Proceed with #1224

MicahZoltu commented:
I have become convinced by Nick that the EIP should be a mechanism for standards only, unrelated to deployments.  In such a scheme, there would be no involvement from the core dev team and no step in the process would depend on them doing something (such as implementing, deploying, scheduling a hard fork, etc.)

EIPs would reference `FORK_BLOCK_NUMBER` rather than `CONSTANTINOPLE_BLOCK_NUMBER` when necessary and the implementation requirement would not need to be completed by any _specific_ client, just any client (e.g., someone's fork of go-ethereum, or pyethereum, or whatever).

I don't know if anyone other than Nick and I have bought into this change or not yet, and I have been waiting for my current EIP PR to get merged before undertaking that change (since unlike my current PR, it's _intent_ is to change the process).

fulldecent commented:
@MicahZoltu I have updated this PR consistent with your PR that was merged. Specifically I updated Core Dev -> Client Dev.

The goal of this PR is to make README.md reflect with our new reality.

There are always more improvements to make. But I believe this PR is mergeable for now.

MicahZoltu commented on README.md:
```diff
@@ -1,34 +1,26 @@
-Ethereum Improvement Proposals (EIPs) describe standards for the Ethereum platform, including core protocol specifications, client APIs, and contract standards.
+# Ethereum Improvement Proposals (EIPs) [![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/ethereum/EIPs?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)
 
-A browsable version of all current and draft EIPs can be found on [the official EIP site](http://eips.ethereum.org/).
```
Did you intend to remove the link to the website?  I personally have found it valuable on multiple occasions.

fulldecent replied:
Added back in c6c1acf76328f8632473c8bed1ef08e8f688d2ae

fulldecent replied:
I believe this one is resolved

MicahZoltu commented on README.md:
```diff
@@ -1,34 +1,26 @@
-# EIPs [![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/ethereum/EIPs?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)
-Ethereum Improvement Proposals (EIPs) describe standards for the Ethereum platform, including core protocol specifications, client APIs, and contract standards.
```
This should probably be kept. It defines what EIP stands for (which is then used throughout the document) and it gives a pithy introduction.  Did you have a reason for removing it?

fulldecent replied:
Thank you, added back in fbe543c79627260d5c10ab8cbd40abca693a03bd

MicahZoltu commented on README.md:
```diff
@@ -1,34 +1,26 @@
+# Ethereum Improvement Proposals (EIPs) [![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/ethereum/EIPs?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)
 
-A browsable version of all current and draft EIPs can be found on [the official EIP site](http://eips.ethereum.org/).
+**Before you initiate a pull request**, please read the process document. Ideas should be thoroughly discussed on the [Ethereum Magicians forums](https://ethereum-magicians.org) first.
```
I like Ethereum magicians, but I know some people don't like it.  In the past, Gitter and GitHub issues have also been recommended for discussion prior to an EIP.  Any reason not to include them here?

fulldecent replied:
I myself don't like Ethereum Magicians. Only because email notifications aren't working.

But most importantly, we should recommend a single place where people can go to have in depth discussions about standards and improvements. And GitHub is not a forum so GitHub issues is probably not the correct place.

To be consistent, this issue also directly the repository owner to disable the Github Issues feature for this repository. Such is consistent with how Apple's Swift Evolution repository is governed. And I have borrowed ideas heavily from them -- including the two-week EIP review.

MicahZoltu commented on README.md:
```diff
@@ -1,34 +1,26 @@
 
- - **For a Standards Track EIP of type Core**, ask to have your issue added to [the agenda of an upcoming All Core Devs meeting](https://github.com/ethereum/pm/issues), where it can be discussed for inclusion in a future hard fork. If implementers agree to include it, the EIP editors will update the state of your EIP to 'Accepted'.
- - **For all other EIPs**, open a PR changing the state of your EIP to 'Final'. An editor will review your draft and ask if anyone objects to its being finalised. If the editor decides there is no rough consensus - for instance, because contributors point out significant issues with the EIP - they may close the PR and request that you fix the issues in the draft before trying again.
+- [Consensus Changes](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1.md#eip-types) require people running an Ethereum client to update their software to implement the change. If enough people do not upgrade then an existing network will fork into two separate networks.
```
Links should be relative, not absolute.  That way they follow in forks (e.g., when creating a PR).

fulldecent replied:
Thank you, fixed in 879ebdd9e45bb9d874d053f673230ed8f0ccd849

fulldecent commented:
@MicahZoltu I have responded to each of your comments. May you please continue your review?

fulldecent commented:
@MicahZoltu ping

MicahZoltu commented on README.md:
```diff
@@ -1,34 +1,30 @@
 Ethereum Improvement Proposals (EIPs) describe standards for the Ethereum platform, including core protocol specifications, client APIs, and contract standards.
 
-A browsable version of all current and draft EIPs can be found on [the official EIP site](http://eips.ethereum.org/).
+**Before you initiate a pull request**, please read the process document. Ideas should be thoroughly discussed on the [Ethereum Magicians forums](https://ethereum-magicians.org) first.
```
There has been pushback in the past about "forcing" people to use ethereum-magicians forum.  Consider weakening the language here to something like:
> Before you initiate a pull request, please read the process document.  Ideas should be thoroughly discussed prior to opening a pull request, such as on the Ethereum Magicians forums or in a GitHub issue in this repository.

fulldecent replied:
Addressing the extreme quantity of low quality issues in this repo is still an issue. (Solution: close GitHub issues and refer to Ethereum Magicians.) But if that will slow this PR down, then I am happy to do it separately.

Updated in 6abedf8

fulldecent replied:
Does this update address the issue?

MicahZoltu commented on README.md:
```diff
@@ -1,34 +1,30 @@
+- [Draft](https://eips.ethereum.org/all) proposals which intend to complete the EIP review process.
+- [Last Calls](https://eips.ethereum.org/all) for proposals that may become final (see also [RSS feed](https://eips.ethereum.org/last-call.xml)).
+- [Accepted](https://eips.ethereum.org/all) and [Deferred](https://eips.ethereum.org/all) proposals which are awaiting implementation or deployment by Ethereum client developers.
+- [Final](https://eips.ethereum.org/all) and [Active](https://eips.ethereum.org/all) proposals that are recorded.
```
All of these links go to the same place.  From a UX point of view, having multiple links go to the same place is confusing because users tend to assume that each link is unique.  If we only have one link to provide, then consider putting it up in the leading sentence:
> This repository tracks the ongoing status of [EIPs](https://eips.ethereum.org/all)

fulldecent replied:
I am just raising the expectations on the index.md. I'll go ahead and and edit that.

Updated in 6abedf8 and 477a314

MicahZoltu commented on README.md:
```diff
@@ -1,34 +1,30 @@
-* **Final (Core)** - an EIP that the Core Devs have decide to implement and release in a future hard fork or has already been released in a hard fork
-* **Deferred** - an EIP that is not being considered for immediate adoption. May be reconsidered in the future for a subsequent hard fork.
+- [Consensus Changes](EIPS/eip-1.md#eip-types) require people running an Ethereum client to update their software to implement the change. If enough people do not upgrade then an existing network will fork into two separate networks.
+- [Other Proposals](EIPS/eip-1.md#eip-types) do not require a change in client software and can be used by any interested party.
```
Same comment as above, both these links go to the same place, despite hinting that they go to different places.  Consider moving the leak to the lead-in sentence:
> ... proposals in two [categories](EIPS/eip-1.md#eip-types):

fulldecent replied:
I have higher expectations for a future EIP-1. But unlike change the all.html, I will defer editing EIP-1 at this time. 

I request a slight exception to the UX rule in this case.

fulldecent commented:
Hello @Arachnid, may I please ask for your review on this version?

Arachnid commented on README.md:
```diff
@@ -1,34 +1,30 @@
+## Project Goal
 
-When you believe your EIP is mature and ready to progress past the draft phase, you should do one of two things:
+The Ethereum Improvement Proposals repository exists as a place to discuss concrete proposals with potential users of the proposal and the Ethereum community at large.
```
No it doesn't; discussion is explicitly intended to take place elsewhere.

fulldecent replied:
Great, updated in ddec7ba

Arachnid commented on README.md:
```diff
@@ -1,34 +1,30 @@
-# Contributing
+This repository tracks the ongoing status of EIPs. It contains:
 
- 1. Review [EIP-1](EIPS/eip-1.md).
```
Why have you deleted all this explanatory text?

fulldecent replied:
Currently the README.md is focused on telling people how to dump whatever questions or ideas are in their head as quickly as possible into an EIP, which is a proposal to tell everybody else about a change they should make.

The proposed README.md focuses instead on explaining what an EIP is and how people should be more disciplined before they think about starting one.

// Ideas stolen from https://github.com/apple/swift-evolution which has a very good change management process.

Arachnid commented on _includes/eiptable.html:
```diff
@@ -11,7 +11,7 @@
   {% assign count = eips|size %}
   {% if count > 0 %}
-    <h2>{{status}}</h2>
+    <h2 id="{{status|slugify}}">{{status}}</h2>
```
Why?

axic replied:
Nice, can we merge this line at least? :)

axic replied:
Moved to #2277.

fulldecent commented:
Here is an example of yet another person that came to the EIP repository to use it as a forum (that's bad) rather than a place to work on actual standards development to promote interoperability among actual products that are or will be deployed (that's good).

➡️  https://github.com/ethereum/EIPs/issues/1621

You can confirm by this focusing statement in that draft EIP:

> "I would like to start a discussion here for the viability and the merit of the proposal as well as design." 

---

I believe #1621 is wildly out of scope for this repository at this stage and wastes time for all parties involved. Instead, such exploratory discussion should be OUTSIDE of a standards setting like this. It should be on Reddit or wherever else.

We should make this clear to people when they first visit this repository. We can do this easily. Just merge this PR.

axic commented on README.md:
```diff
@@ -1,34 +1,30 @@
 Ethereum Improvement Proposals (EIPs) describe standards for the Ethereum platform, including core protocol specifications, client APIs, and contract standards.
 
-A browsable version of all current and draft EIPs can be found on [the official EIP site](http://eips.ethereum.org/).
+**Before you initiate a pull request**, please read the process document. Ideas should be thoroughly discussed prior to opening a pull request, such as on the [Ethereum Magicians forums](https://ethereum-magicians.org) or in a GitHub issue in this repository.
```
The second sentence is now emphasized in EIP-1. Perhaps not needed here.

fulldecent replied:
Good call, addressed in 90f04bd

axic commented on README.md:
```diff
@@ -1,34 +1,30 @@
 Ethereum Improvement Proposals (EIPs) describe standards for the Ethereum platform, including core protocol specifications, client APIs, and contract standards.
 
-A browsable version of all current and draft EIPs can be found on [the official EIP site](http://eips.ethereum.org/).
+**Before you initiate a pull request**, please read the process document. Ideas should be thoroughly discussed prior to opening a pull request, such as on the [Ethereum Magicians forums](https://ethereum-magicians.org) or in a GitHub issue in this repository.
```
```suggestion
**Before you initiate a pull request**, please read the [EIP-1](https://eips.ethereum.org/EIPS/eip-1) process document. Ideas should be thoroughly discussed prior to opening a pull request, such as on the [Ethereum Magicians forums](https://ethereum-magicians.org) or in a GitHub issue in this repository.
```

fulldecent replied:
Thank you. Added in f8036cc. Superseding the commit above.

axic commented:
@fulldecent @Arachnid what is the status of this PR?

fulldecent commented:
@axic Thank you. The status is: just rebased against master, build test passing, ready to merge, attracting consensus to merge

axic commented on README.md:
```diff
@@ -1,35 +1,31 @@
-Ethereum Improvement Proposals (EIPs) describe standards for the Ethereum platform, including core protocol specifications, client APIs, and contract standards.
+# Ethereum Improvement Proposals (EIPs) [![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/ethereum/EIPs?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)
 
-A browsable version of all current and draft EIPs can be found on [the official EIP site](https://eips.ethereum.org/).
```
I think this line is useful here. Especially if you want to make the distinction between consuming EIPs (the eips.ethereum.org website) and submitting EIPs (the readme).

fulldecent replied:
This line is there but it is moved below the warning about opening new PRs. The motivation of course being that people are opening PRs with reckless abandon whereas nobody is having any trouble finding an EIP if they wanted to.

fulldecent commented:
@axic Thank you, I appreciate your updates to keep this up to date

axic commented:
@fulldecent can you fix the conflicts please?

fulldecent commented:
@axic resolved

axic commented on README.md:
```diff
@@ -4,35 +4,32 @@
-Your first PR should be a first draft of the final EIP. It must meet the formatting criteria enforced by the build (largely, correct metadata in the header). An editor will manually review the first PR for a new EIP and assign it a number before merging it. Make sure you include a `discussions-to` header with the URL to a discussion forum or open GitHub issue where people can discuss the EIP as a ...
+Achieving "Final" status in this repository only represents that a proposal has been reviewed for technical accuracy. It is solely the responsibilty of the implementer to decide if these proposals will be useful to you and if they represent best practice.
 
-If your EIP requires images, the image files should be included in a subdirectory of the `assets` folder for that EIP as follows: `assets/eip-N` (where **N** is to be replaced with the EIP number). When linking to an image in the EIP, use relative links such as `../assets/eip-1/image.png`.
```
This information is part of EIP-1.

fulldecent replied:
Agreed. Removing from REAMDE

axic commented on README.md:
```diff
@@ -4,35 +4,32 @@
+- [Final](https://eips.ethereum.org/all#final) and [Active](https://eips.ethereum.org/all#active) proposals that are recorded.
+- The [EIP process](EIPS/eip-1.md#eip-work-flow) that governs the EIP repository.
 
-Your first PR should be a first draft of the final EIP. It must meet the formatting criteria enforced by the build (largely, correct metadata in the header). An editor will manually review the first PR for a new EIP and assign it a number before merging it. Make sure you include a `discussions-to` header with the URL to a discussion forum or open GitHub issue where people can discuss the EIP as a ...
```
This is not highlighted in index.html nor EIP-1. I think ti should be added.

axic commented on README.md:
```diff
@@ -4,35 +4,32 @@
-When you believe your EIP is mature and ready to progress past the draft phase, you should do one of two things:
+## Project Goal
 
- - **For a Standards Track EIP of type Core**, ask to have your issue added to [the agenda of an upcoming All Core Devs meeting](https://github.com/ethereum/pm/issues), where it can be discussed for inclusion in a future hard fork. If implementers agree to include it, the EIP editors will update the state of your EIP to 'Accepted'.
```
This is not highlighted in index.html nor EIP-1. I think it should be added.

fulldecent replied:
Specifications for the process should be in EIP-1. This information should be out of scope for the README.

axic commented on README.md:
```diff
@@ -4,35 +4,32 @@
- - **For all other EIPs**, open a PR changing the state of your EIP to 'Final'. An editor will review your draft and ask if anyone objects to its being finalised. If the editor decides there is no rough consensus - for instance, because contributors point out significant issues with the EIP - they may close the PR and request that you fix the issues in the draft before trying again.
+The Ethereum Improvement Proposals repository exists as a place to share concrete proposals with potential users of the proposal and the Ethereum community at large.
 
-# EIP Status Terms
```
These are part of index.html and EIP-1.

MicahZoltu commented on README.md:
```diff
@@ -4,35 +4,32 @@
-* **Accepted** - a core EIP that has been in Last Call for at least 2 weeks and any technical changes that were requested have been addressed by the author. The process for Core Devs to decide whether to encode an EIP into their clients as part of a hard fork is not part of the EIP process. If such a decision is made, the EIP will move to final.
-* **Final (non-Core)** - an EIP that has been in Last Call for at least 2 weeks and any technical changes that were requested have been addressed by the author.
-* **Final (Core)** - an EIP that the Core Devs have decided to implement and release in a future hard fork or has already been released in a hard fork. 
+- [Consensus Changes](EIPS/eip-1.md#eip-types) require people running an Ethereum client to update their software to implement the change. If enough people do not upgrade then an existing network will fork into two separate networks.
+- [Other Proposals](EIPS/eip-1.md#eip-types) do not require a change in client software and can be used by any interested party.
```
What value do we get in making this distinction?  Why is this distinction more valuable than the current Core, Network, Interface, ERC?

fulldecent replied:
I believe the audience for README.md will be the wide community of Ethereum users. This includes people that will be coming by (driving by?) to make their opinions heard on the likes of the next contentious hard fork, as well as others that are implementers/developers.

I'm guessing that anybody who has more than one Ether for more than one or two years will have a strong opinion on The DAO fork, and maybe even the latest fork. But I'll bet that less than 50% would know if those were Core, Network, Interface, or ERC EIPs.

Therefore I have added this distinction in plain language so that the intended audience can easily understand what they are looking at.

MicahZoltu replied:
I suspect that the number of people in the intersection of "drive by brigadier" and "stopped to read the Readme" and "didn't comprehend the categories" is small enough that it isn't worth adding complexity (another layer of abstraction) to the EIP process, EIPs page, etc.

fulldecent replied:
We disagree on the target audience.

But we agree that this PR can go on without that part.

MicahZoltu replied:
If you want to extract these two lines out into a separate PR we can merge this without being blocked by this discussion, and then we can have a more focused discussion on just this topic (with hopefully more people getting involved).  I'm 👍 to merge everything else, but would like to discuss these lines in particular more before merging.

MicahZoltu replied:
Note: I'm OK with the deletion above these two lines being part of this PR, even if these two lines aren't included in this PR.  We would want to remove line 27 (of your branch) as well and pull it out to the new PR if we decide to go that route.  So the new PR would only add the following, and this PR would remove the status description section (which really should be in EIP-1 only anyway).
> We promote high-quality, peer-reviewed proposals in two categories:
> - [Consensus Changes](EIPS/eip-1.md#eip-types) require people running an Ethereum client to update their software to implement the change. If enough people do not upgrade then an existing network will fork into two separate networks.
> - [Other Proposals](EIPS/eip-1.md#eip-types) do not require a change in client software and can be used by any interested party.

fulldecent replied:
Fixed at 922a1abe861dde67abf4eda9486e380be49c9968

MicahZoltu reviewed (commented):
2 years later and now an editor (rather than just an annoying commenter on the internet), I generally approve of this change, though I left a few notes that may be worth discussing/addressing before merge.  Leaving this as a Comment rather than a Reject or Approve since I don't feel super strongly about any of them.  The comment about Consensus/Other is the only one that really makes me hesitant to Approve this as "strictly better than what was here before".

fulldecent commented on README.md:
```diff
@@ -4,35 +4,32 @@
+- The [EIP process](EIPS/eip-1.md#eip-work-flow) that governs the EIP repository.
 
-Your first PR should be a first draft of the final EIP. It must meet the formatting criteria enforced by the build (largely, correct metadata in the header). An editor will manually review the first PR for a new EIP and assign it a number before merging it. Make sure you include a `discussions-to` header with the URL to a discussion forum or open GitHub issue where people can discuss the EIP as a ...
+Achieving "Final" status in this repository only represents that a proposal has been reviewed for technical accuracy. It is solely the responsibility of the implementer to decide if these proposals will be useful to you and if they represent best practice.
```
Perfect.

fulldecent commented on README.md:
```diff
@@ -4,35 +4,32 @@
- 3. Add your EIP to your fork of the repository. There is a [template EIP here](eip-template.md).
- 4. Submit a Pull Request to Ethereum's [EIPs repository](https://github.com/ethereum/EIPs).
+- [Draft](https://eips.ethereum.org/all#draft) proposals which intend to complete the EIP review process.
+- [Last Calls](https://eips.ethereum.org/all#last-call) for proposals that may become final (see also [RSS feed](https://eips.ethereum.org/last-call.xml)).
```
LGTM

fulldecent commented:
@MicahZoltu thanks for the feedback, all your items replied and/or merged

MicahZoltu reviewed (approved):
Since this is a change to the Readme, I would like to get an approval from at least one other active editor, but I'm 👍 on this.  @lightclient @axic do either of you have thoughts on these changes?

If we believe these changes are iteratively better than what is currently in the Readme, I propose we merge them and we can then iterate further afterward.  If we think this is a step backward, then we should discuss further before merging.

Souptacular reviewed (approved):
LGTM. Great work simplifying this.

lightclient reviewed (approved):
This looks good. The `validation` section is out-of-date, but I can fix that in a separate PR.

MicahZoltu commented:
Given this has been open for over 2 years I'm satisfied with 3 approvals.  Going to merge this, we can always iterate on it later if other people have feedback.

fulldecent commented:
Cool, thank you
